### PR TITLE
Token blacklisting

### PR DESCRIPTION
**Pull request by mzugcic**
This PR:
- enables users to **blacklist** their own tokens to prevent login
- login with blacklisted tokens is disabled in the authentication

### CODE REVIEW

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    except User.DoesNotExist:
+        msg = _("User doesn't exist.")
+        raise serializers.ValidationError(msg)
+
+    if not user.is_active:
+        msg = _('User account is disabled.')
+        raise serializers.ValidationError(msg)
+
+    return user
+
```
Comment (mzugcic) [inline src/rest_framework_jwt/blacklist/serializers.py]: `_check_payload` & `_check_user` should maybe be changed to public functions so they can be used here if needed without copying them. We had a discussion about this while you were gone.

What do you think @fitodic ?

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+import jwt
+
+from django.contrib.auth import get_user_model
+from django.utils.translation import ugettext as _
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/serializers.py]: `ugettext` has entered a deprecation period and will raise deprecation warnings. I suggest handling this via the `compat` module.

File: src/rest_framework_jwt/blacklist/serializers.py (hunk 1 above)
Comment (fitodic) [reply src/rest_framework_jwt/blacklist/serializers.py]: Sure, go ahead and expose them. Better yet, please move them to the `utils` module.

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    return user
+
+
+class BlacklistTokenSerializer(serializers.Serializer):
+    """
+    Serializer used for blacklisting tokens.
+    """
+
+    token = serializers.CharField(required=False)
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/serializers.py]: Why is this not required?

File: src/rest_framework_jwt/blacklist/views.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+from rest_framework.viewsets import ModelViewSet
+
+from rest_framework_jwt.blacklist.models import BlacklistedToken
+from .serializers import BlacklistTokenSerializer
+from ..authentication import JSONWebTokenAuthentication
+
+
+class BlacklistView(ModelViewSet):
+    permission_classes = (IsAuthenticated, )
+    authentication_classes = (JSONWebTokenAuthentication, )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: The [documentation](https://github.com/Styria-Digital/django-rest-framework-jwt/blob/master/docs/index.md#usage) states the authentication class should be defined in the `REST_FRAMEWORK` settings to make it globally available. This renders this assignment redundant at best, and difficult to debug at worst (if someone uses a customized `JSONWebTokenAuthentication` class).

File: src/rest_framework_jwt/blacklist/views.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            return Response(
+                serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
+
+        token = serializer.data.get('token')
+        token_payload = JSONWebTokenAuthentication.jwt_decode_token(token)
+        token_username = JSONWebTokenAuthentication.jwt_get_username_from_payload(token_payload)
+
+        if token_username != request.user.username:
+            return Response(status=status.HTTP_403_FORBIDDEN)
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: I suggest handling this case using [object-level permissions](https://www.django-rest-framework.org/api-guide/permissions/#object-level-permissions). You'll need to create a separate permission class, but that would enable users of this library to easily customize the permission handling process.

File: src/rest_framework_jwt/blacklist/views.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    authentication_classes = (JSONWebTokenAuthentication, )
+    queryset = BlacklistedToken.objects.all().order_by('-blacklisted_at')
+    serializer_class = BlacklistTokenSerializer
+
+    def create(self, request, *args, **kwargs):
+        serializer = self.get_serializer(data=request.data)
+        if not serializer.is_valid():
+            return Response(
+                serializer.errors, status=status.HTTP_400_BAD_REQUEST
+            )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Isn't `serializer.is_valid(raise_exception=True)` enough?

File: src/rest_framework_jwt/blacklist/views.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+        token = serializer.data.get('token')
+        token_payload = JSONWebTokenAuthentication.jwt_decode_token(token)
+        token_username = JSONWebTokenAuthentication.jwt_get_username_from_payload(token_payload)
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Since the token is not required by the serializer, what happens if the `token` is `None`?

File: src/rest_framework_jwt/blacklist/views.py [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+        if token_username != request.user.username:
+            return Response(status=status.HTTP_403_FORBIDDEN)
+
+        blacklisted_token = BlacklistedToken.objects.get_or_create(token=token)[0]
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: 1. When using `get_or_create`, if you don't need the `created` value, a common workaround is to assign it to [an underscore "throwaway" variable](https://stackoverflow.com/a/5893946): `_`. This indicates that this value is intentionally ignored.

2. This statement is susceptible to race conditions (as mentioned in [`models.py`](https://github.com/Styria-Digital/django-rest-framework-jwt/pull/27/files#diff-a7aff25f8face4166d47a31b3845918bR5)).

File: tests/views/test_blacklist.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+from rest_framework.reverse import reverse
+
+from rest_framework_jwt.blacklist.models import BlacklistedToken
+from rest_framework_jwt.serializers import JSONWebTokenAuthentication
+
+
+def test_only_user_can_blacklist_own_token(
+    user, staff_user, create_authenticated_client
+):
+    assert not BlacklistedToken.objects.exists()
```
Comment (fitodic) [inline tests/views/test_blacklist.py]: Each test has a clean slate (or should have, but in this library it _has_), which means that this assertion is redundant.

File: tests/views/test_blacklist.py [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    url = reverse('blacklist:blacklist-list')
+    data = {'token': JSONWebTokenAuthentication.jwt_encode_payload(payload)}
+
+    # Response should return status 403 Forbidden if anyone besides the
+    # owner tries to blacklist a token
+    response = api_client_staff_user.post(url, data)
+
+    assert response.status_code == status.HTTP_403_FORBIDDEN
+    assert not BlacklistedToken.objects.exists()
```
Comment (fitodic) [inline tests/views/test_blacklist.py]: This is one test case, the one concerning the admin user's permissions.

File: tests/views/test_blacklist.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+    # Owner can blacklist own token
+    api_client = create_authenticated_client(user)
+    response = api_client.post(url, data)
+
+    assert response.status_code == status.HTTP_200_OK
+    assert BlacklistedToken.objects.exists()
```
Comment (fitodic) [inline tests/views/test_blacklist.py]: This is another test case, the one concerning the token owner's test case.

File: src/rest_framework_jwt/blacklist/views.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        response = Response(
+            JSONWebTokenAuthentication.jwt_create_response_payload(
+                blacklisted_token.token, request.user, request
+            )
+        )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Commonly, the response status code for successful POST requests is [201](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201).

However, this implementation only allows the user to blacklist its own token, meaning that the user is effectively logged out. In that case, a [205](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/205) response a more appropriate which tells the client to refresh the page.

File: tests/conftest.py [hunk 13]
```diff
@@ -52,6 +50,30 @@ def user(create_user):
...
+@pytest.fixture
+def super_user(create_user):
+    return create_user(
+        username="superusername",
+        email="superusername@example.com",
+        password="super",
+        is_active=True,
+        is_staff=True,
+        is_superuser=True,
+    )
```
Comment (fitodic) [inline tests/conftest.py]: Why is this fixture added if it is not used?

File: tests/views/test_blacklist.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
+# -*- coding: utf-8 -*-
```
Comment (fitodic) [inline tests/views/test_blacklist.py]: I can think of a couple of more test cases:

* The superuser tries to blacklist the token
* `rest_framework_jwt.blacklist` is not registered in the `INSTALLED_APPS`
* User A tries to blacklist user B
* A user tries to blacklist a token that is already blacklisted.

File: src/rest_framework_jwt/blacklist/views.py (hunk 12 above)
Comment (mzugcic) [reply src/rest_framework_jwt/blacklist/views.py]: All successful POST requests return response status code **201** because they create a Blacklisted token.

File: tests/views/test_blacklist.py (hunk 14 above)
Comment (mzugcic) [reply tests/views/test_blacklist.py]: All suggested test cases, except the second bullet (**blacklist** app not registered in `INSTALLED_APPS`), have been added.

File: src/rest_framework_jwt/blacklist/admin.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin
+from django.utils.translation import gettext_lazy as _
+
+from .models import BlacklistedToken
+
+
+@admin.register(BlacklistedToken)
+class BlacklistedTokenAdmin(admin.ModelAdmin):
+    list_display = ('token', )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/admin.py]: What about the other fields, such as `user` and `blacklisted_by`?

A temporal filter would be useful as well.

File: src/rest_framework_jwt/authentication.py [hunk 16]
```diff
@@ -58,6 +61,10 @@ def authenticate(self, request):
         if jwt_value is None:
             return None
 
+        if BlacklistedToken.objects.filter(token=jwt_value).exists():
```
Comment (fitodic) [inline src/rest_framework_jwt/authentication.py]: Just a reminder to check the cookies as well.

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 17]
```diff
@@ -0,0 +1,26 @@
+from django.contrib.auth import get_user_model
+
+from rest_framework.permissions import IsAuthenticated
+
+from rest_framework_jwt.authentication import JSONWebTokenAuthentication
+from rest_framework_jwt.utils import check_user
+
+
+class CanBlacklist(IsAuthenticated):
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: Where is the `IsAuthenticated` check applied? If I'm reading this correctly, the current state of things will raise an `AttributeError` on [line 16](https://github.com/Styria-Digital/django-rest-framework-jwt/pull/27/files#diff-e17925b4119cfea3a2bf30816c188800R16) if the user is anonymous. This results in a 500 response to the client, not 401.

A test case for anonymous users would also help. :slightly_smiling_face:

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+        user_id = request.data.get('user')
+
+        if not user_id:
+            user_id = request.user.id
+
+        user = get_user_model().objects.get(id=user_id)
+        payload = JSONWebTokenAuthentication.jwt_create_payload(user)
+        user = check_user(payload)
+
+        if user.username != request.user.username:
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: The `username` is fine, but `pk`s never change.

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+            if not request.user.is_staff or user.is_superuser:
+                return False
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: [Object level permissions](https://www.django-rest-framework.org/api-guide/permissions/#object-level-permissions) ([`has_object_permission`](https://www.django-rest-framework.org/api-guide/permissions/#custom-permissions)) are more appropriate for these types of checks.

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+        fields = ('user', )
+
+    def validate(self, data):
+        user = data.get('user')
+
+        payload = JSONWebTokenAuthentication.jwt_create_payload(user)
+        token = JSONWebTokenAuthentication.jwt_encode_payload(payload)
+
+        iat = payload.get('iat', unix_epoch())
+        expires_at_unix_time = iat + api_settings.JWT_EXPIRATION_DELTA.total_seconds()
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/serializers.py]: Are you creating a new token based on the `user` instead of blacklisting the existing from the request? If so, is the expiration time the same? Or better yet, is it even the same token?

File: src/rest_framework_jwt/blacklist/views.py [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+from rest_framework_jwt.blacklist.models import BlacklistedToken
+
+from ..authentication import JSONWebTokenAuthentication
+from .permissions import CanBlacklist
+from .serializers import BlacklistTokenSerializer
+
+
+class BlacklistView(ModelViewSet):
+    permission_classes = (CanBlacklist, )
+    authentication_classes = (JSONWebTokenAuthentication, )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Is this authentication necessary? What happens if someone overrides it and sets it in the `JWT_AUTH` settings?

File: src/rest_framework_jwt/blacklist/views.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    queryset = BlacklistedToken.objects.all().order_by('-blacklisted_at')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Why is this `order_by` necessary?

File: src/rest_framework_jwt/blacklist/views.py [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+    serializer_class = BlacklistTokenSerializer
+
+    def create(self, request, *args, **kwargs):
+        try:
+            if 'user' not in request.data:
+                # This case should occur only when this view is used at logout.
+                request.data.update({'user': request.user.id})
+            return super(BlacklistView, self).create(request, *args, **kwargs)
+        except IntegrityError:
+            raise ValidationError('User\'s token has already been blacklisted.')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: `perform_create` is the place where this should be applied.

File: src/rest_framework_jwt/blacklist/views.py [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+    authentication_classes = (JSONWebTokenAuthentication, )
+    queryset = BlacklistedToken.objects.all().order_by('-blacklisted_at')
+    serializer_class = BlacklistTokenSerializer
+
+    def create(self, request, *args, **kwargs):
+        try:
+            if 'user' not in request.data:
+                # This case should occur only when this view is used at logout.
+                request.data.update({'user': request.user.id})
+            return super(BlacklistView, self).create(request, *args, **kwargs)
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: Why is the `user` the serializer's only/obligatory field? Are we blacklisting tokens, or users, or both?

File: src/rest_framework_jwt/blacklist/signals.py [hunk 25]
```diff
@@ -0,0 +1,9 @@
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from .models import BlacklistedToken
+
+
+@receiver(post_save, sender=BlacklistedToken)
+def delete_stale_tokens(sender, instance, **kwargs):
+    BlacklistedToken.objects.delete_stale_tokens()
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/signals.py]: Is this signal registered anywhere? Or tested?

File: changelog.d/27.feature.md [hunk 26]
```diff
@@ -0,0 +1 @@
+Blacklisting allows the service to blacklist user's own token from HTTP header or cookies
```
Comment (fitodic) [inline changelog.d/27.feature.md]: Blacklisting allows the user to blacklist its own token.

File: changelog.d/27.removal.md [hunk 27]
```diff
@@ -0,0 +1 @@
+Removed occurrences of Python 3.3 & 3.4 from build configuration files since they were deprecated from Ubuntu
```
Comment (fitodic) [inline changelog.d/27.removal.md]: Drop support for Python 3.3 and 3.4

File: docs/index.md [hunk 28]
```diff
@@ -135,6 +135,10 @@ Impersonation allows the service to perform actions on the client’s behalf. A
 
 By default, only superusers (`user.is_superuser == True`) can impersonate other accounts. If you need to customize the permission handling process, override the `ImpersonateJSONWebTokenView`'s [`permission_classes` attribute](https://www.django-rest-framework.org/api-guide/permissions/#setting-the-permission-policy). 
 
+## Blacklisting Tokens
+
+Blacklisting allows users to blacklist their own token from the HTTP header or cookies by redirecting to the `BlacklistView` in which case the view is serves as a logout.
```
Comment (fitodic) [inline docs/index.md]: How are users redirected to the `BlacklistView`?

File: src/rest_framework_jwt/blacklist/admin.py [hunk 29]
```diff
@@ -0,0 +1,9 @@
+from django.contrib import admin
+
+from .models import BlacklistedToken
+
+
+@admin.register(BlacklistedToken)
+class BlacklistedTokenAdmin(admin.ModelAdmin):
+    list_display = ('user', 'token', 'expires_at', )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/admin.py]: What about `blacklisted_at`? Should this field be added as well?

File: src/rest_framework_jwt/blacklist/models.py [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+
+class BlacklistedTokenManager(models.Manager):
+    def delete_stale_tokens(self):
+        return self.filter(expires_at__lt=datetime.utcnow()).delete()
+
+
+class BlacklistedToken(models.Model):
+    token = models.TextField(db_index=True, unique=True)
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name='owner')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/models.py]: Blacklisted tokens are accessed via `user.owner`?

File: src/rest_framework_jwt/blacklist/models.py [hunk 31]
```diff
@@ -0,0 +1,22 @@
...
+    expires_at = models.DateTimeField(db_index=True)
+    blacklisted_at = models.DateTimeField(auto_now_add=True)
+
+    objects = BlacklistedTokenManager()
+
+    def __str__(self):
+        return 'Blacklisted token - {} - {}'.format(self.user, self.token)
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/models.py]: Is the token necessary in the output?

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 32]
```diff
@@ -0,0 +1,20 @@
+
+from rest_framework.permissions import BasePermission
+
+from rest_framework_jwt.blacklist.models import BlacklistedToken
+from rest_framework_jwt.settings import api_settings
+
+
+class IsAuthenticatedAndNotBlacklisted(BasePermission):
+    message = 'You are not authenticated or have been blacklisted.'
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: Isn't this permission doing too much? What happens if users of this library want to combine the `blacklist` permission with their customized `IsAuthenticated` permissions?

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+
+    def has_permission(self, request, view):
+        if request.user and not request.user.is_authenticated:
+            return False
+
+        if api_settings.JWT_AUTH_COOKIE:
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: Is the impersonation cookie intentionally omitted?

File: src/rest_framework_jwt/blacklist/permissions.py [hunk 34]
```diff
@@ -0,0 +1,20 @@
...
+            token = request.COOKIES.get(api_settings.JWT_AUTH_COOKIE)
+        else:
+            token = request.auth.decode('utf-8')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/permissions.py]: Django's `force_text` or `force_str` would be more appropriate in this scenario.

File: src/rest_framework_jwt/blacklist/views.py [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+                imp_user_token = request.COOKIES.get(api_settings.JWT_IMPERSONATION_COOKIE)
+                if imp_user_token:
+                    return imp_user_token
+
+            if api_settings.JWT_AUTH_COOKIE:
+                return request.COOKIES.get(api_settings.JWT_AUTH_COOKIE)
+
+        # Using decode because the token from the header is 'bytes' type and raises
+        # error on serializer's `is_valid()` method
+        return auth[1].decode('utf-8')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: `force_text` or `force_str` would be more appropriate if you want to deal with strings.

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+
+        return {
+            'token':
+                token,
+            'user':
+                check_user(payload),
+            'expires_at':
+                make_aware(datetime.utcfromtimestamp(expires_at_unix_time)),
+        }
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/serializers.py]: Why is this necessary? Is the output used anywhere or is it just for validation purposes?

File: src/rest_framework_jwt/blacklist/serializers.py [hunk 37]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class BlacklistTokenSerializer(serializers.ModelSerializer):
+    """
+    Serializer used for blacklisting tokens.
+    """
+
+    class Meta:
+        model = BlacklistedToken
+        fields = ('token', )
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/serializers.py]: You can specify that the token is optional via [`extra_kwargs`](https://www.django-rest-framework.org/api-guide/serializers/#additional-keyword-arguments). This avoids any confusion when users are looking at automatically generated documentation (e.g. Swagger).

File: src/rest_framework_jwt/blacklist/views.py [hunk 38]
```diff
@@ -0,0 +1,57 @@
...
+        if 'token' not in request.data:
+            request.data.update({'token': self.get_jwt_value(request)})
+
+        return super(BlacklistView, self).create(request, *args, **kwargs)
+
+    def perform_create(self, serializer):
+        try:
+            serializer.save()
+        except IntegrityError:
+            raise ValidationError('User\'s token has already been blacklisted.')
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/views.py]: What do you say we return a [409](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409) response? 403 oftentimes indicates that the user doesn't have the required permissions to perform a certain action, whereas a 409 indicates that the resource already exists. Either that or a 400, but then again, the server _can_ process the request, it just refuses to on account of the state in the database.

File: tests/views/test_blacklist.py [hunk 39]
```diff
@@ -0,0 +1,124 @@
...
+    response = api_client.post(url, data, format='json')
+
+    assert response.status_code == status.HTTP_200_OK
+
+    url = reverse('blacklist-list')
+
+    response = api_client.post(url)
+
+    assert response.status_code == status.HTTP_201_CREATED
+    assert BlacklistedToken.objects.first().user_id == super_user.id
```
Comment (fitodic) [inline tests/views/test_blacklist.py]: Shouldn't this action blacklist the impersonation cookie instead?

File: src/rest_framework_jwt/blacklist/models.py [hunk 40]
```diff
@@ -0,0 +1,22 @@
+from datetime import datetime
+
+from django.conf import settings
+from django.db import models
+
+
+class BlacklistedTokenManager(models.Manager):
+    def delete_stale_tokens(self):
+        return self.filter(expires_at__lt=datetime.utcnow()).delete()
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/models.py]: Shouldn't this be handled via a setting? For instance, delete all tokens that are older than the specified delta?

File: src/rest_framework_jwt/blacklist/apps.py [hunk 41]
```diff
@@ -0,0 +1,8 @@
+from django.apps import AppConfig
+
+
+class BlacklistedTokenConfig(AppConfig):
+    name = 'rest_framework_jwt.blacklist'
+
+    def ready(self):
+        import rest_framework_jwt.blacklist.signals  # noqa
```
Comment (fitodic) [inline src/rest_framework_jwt/blacklist/apps.py]: How can users disable this signal if necessary?

File: src/rest_framework_jwt/blacklist/models.py (hunk 30 above)
Comment (mzugcic) [reply src/rest_framework_jwt/blacklist/models.py]: Removed the `related_name` part.

File: docs/index.md [hunk 42]
```diff
@@ -368,6 +380,21 @@ This cookie takes precedence over the `JWT_AUTH_COOKIE`. If you have both cookie
...
+  `JWT_STALE_BLACKLISTED_TOKEN_EXPIRATION_TIME`.
+
+Default is `False`.
+
+### JWT_STALE_BLACKLISTED_TOKEN_EXPIRATION_TIME
+
+This is an instance of Python's `datetime.datetime`. Represents the date & time before which expired
+ blacklisted tokens should be deleted.
+
+Default is `timezone.now()`(timezone aware current moment).
```
Comment (fitodic) [inline docs/index.md]: Is it always aware or does it depend on the `USE_TZ` setting?

File: src/rest_framework_jwt/settings.py [hunk 43]
```diff
@@ -43,6 +44,8 @@
     'JWT_AUTH_COOKIE_SECURE': True,
     'JWT_AUTH_COOKIE_SAMESITE': 'Lax',
     'JWT_IMPERSONATION_COOKIE': None,
+    'JWT_DELETE_STALE_BLACKLISTED_TOKENS': False,
+    'JWT_STALE_BLACKLISTED_TOKEN_EXPIRATION_TIME': timezone.now(),
```
Comment (fitodic) [inline src/rest_framework_jwt/settings.py]: This will be evaluated to the server's startup time, not dynamically evaluated. Try using `timezone.now` which sets it to a function instead.

File: src/rest_framework_jwt/blacklist/views.py (hunk 24 above)
Comment (mzugcic) [reply src/rest_framework_jwt/blacklist/views.py]: We are blacklisting tokens, changed in the new commits.

File: docs/index.md (hunk 42 above)
Comment (mzugcic) [reply docs/index.md]: The value depends on the `USE_TZ` setting. Will be repaired in the next commit.

**Comment by nigoroll** [comment]
So JWTs, by design, cannot be revoked - at least not across many backends. User-Agents can voluntarily forget about a token, and we can change the signing secret/key to invalidate existing tokens, but we cannot make valid JWTs vanish.
Given this, I wonder: Why isn't the solution for logout a simple cookie deletion on the client?
